### PR TITLE
feat: `<Badge />` component

### PR DESCRIPTION
**Pull request by alexcarpenter**
🎨 [Figma Component](https://www.figma.com/file/7cYgDM618stjYUHDqAfRec/Components?node-id=211%3A719)
🔍 [Preview Link](https://react-components-git-acbadge-component-hashicorp.vercel.app/components/badge)

---

## Description

This PR aims to introduce a new `<Badge />` component in line with the new Web Pattern Library Figma component implementation.

### PR Checklist 🚀

Items in this checklist may not may not apply to your PR, but please consider each item carefully.

- [x] Add Asana and Preview links above.
- [x] Conduct thorough self-review.
- [x] Add or update tests as appropriate.
- [x] Conduct reasonable cross browser testing for both compatibility and responsive behavior (We have a [Sauce Labs](https://app.saucelabs.com/) account for this, if you don't have access, just ask!).
- [x] Conduct reasonable accessibility review (use the [WAS](https://accessible.org/Web-Accessibility-Standards-WAS-2.pdf) as a guide or an [axe browser plugin](https://www.deque.com/axe/) until we establish more formal checks).
- [x] Identify (in the description above) and document (add Asana tasks on [this board](https://app.asana.com/0/1100423001970639/list)) any technical debt that you're aware of, but are not addressing as part of this PR.

### CODE REVIEW

File: packages/badge/index.tsx
Comment (dstaley) [inline packages/badge/index.tsx:6]: I'm curious if you considered using a prop for the text value instead, like `<Badge title="" />`?

File: packages/badge/index.tsx
Comment (BRKalow) [inline packages/badge/index.tsx:17]: Did we look at using the product meta styles for this? https://github.com/hashicorp/web-platform-packages/tree/main/packages/product-meta

Comment (dstaley) [inline packages/badge/index.tsx:17]: I'll defer to @ashleemboyer, but I vaguely recall badges like this benefiting from some aria props like `role="status"`. That pertains more to badges that are being updated as a result of user action which doesn't seem to be the intent here, but I'm curious if we should go ahead and add those attributes now to support that use case in the future.

**Review by dstaley — APPROVED** [review]
😍

**Review by BRKalow — COMMENTED** [review]
Looks good!

File: packages/badge/style.module.css
Comment (alexcarpenter) [reply packages/badge/style.module.css]: 👍🏼 https://github.com/hashicorp/react-components/pull/512/commits/78db41fe7e90870d8cf37761d2b1766d00f889c5

File: packages/badge/index.tsx
Comment (ashleemboyer) [reply packages/badge/index.tsx:17]: Setting `role="status"` has side effects, like setting `aria-live` to `polite`, so I'd say let's hold off on adding ARIA until we know more about how this component will be used. But my gut take is that if the text in this badge is static / doesn't update while a user is looking at it, then `role="status"` shouldn't be used.

Source: https://www.w3.org/TR/wai-aria/#status

File: packages/badge/index.tsx
Comment (alexcarpenter) [reply packages/badge/index.tsx:6]: I wanted to propose we use children because it naturally follows how you'd use an html element. For example you want to place text inside a paragraph `<p>Test</p>` vs `<p title="Test" />`. I've found it to be confusing the use of text or title between our components. [Button](https://react-components.vercel.app/components/button) expects `title`, where [CallToAction](https://react-components.vercel.app/components/calltoaction) links expect `text`.

Comment (jmfury) [reply packages/badge/index.tsx:6]: +1 - That was what I had in t[his initial iteration on a project](https://hashicorp-www-next-git-jmhomepage-promo-section-hashicorp.vercel.app/swingset/components/badge), I wondered if there was a strong reason _for_ using `children`

File: packages/badge/index.tsx
Comment (alexcarpenter) [reply packages/badge/index.tsx:17]: I did, but the theme support more than our product colors. We'd be loading unnecessary variable declarations that are not documented to use within the Figma component implementation.

Comment (BRKalow) [reply packages/badge/index.tsx:17]: The product meta theme classes are more than likely already in the document though, so it shouldn't be anything net-new. We could  use a combination of the styles from product meta as well as ones which are for the unique themes that `Badge` supports. Which unnecessary variable declarations are we worried about?

File: packages/badge/index.tsx
Comment (EnMod) [reply packages/badge/index.tsx:6]: +100! The "pure HTML" feel of wrapping an element's content is very intuitive, and as a bonus it promotes composition. I dig this!

File: packages/badge/index.tsx
Comment (alexcarpenter) [reply packages/badge/index.tsx]: @BRKalow Here is an update making use of productStyles https://github.com/hashicorp/react-components/pull/512/commits/282709c8e36a8166ca5df4627c8f8d4fa4689187

Comment (BRKalow) [reply packages/badge/index.tsx]: oh snap

**Review by BRKalow — APPROVED** [review]
👏

**Pull request merged**
